More improvements to cheat sheet

gnprice opened the pull request:
I made a revision pass through the whole document, and made several
kinds of changes:

* Organize it all in hopes of making things easier to find with
  thematic groupings, headers, spacing, and putting the key words at
  the start of each description.

* Give the descriptions more of a common style.

* Explain a few things like `reveal_type` and `Sequence` a little more
  fully.

* Cut the redundant `Tuple` example.

* Add descriptions for the class-definition example.

* Add a description for the generator-function example, and change its
  return-type annotation from `Iterator` to `Iterable`.  That gets us
  out of trying to explain the difference between those, as `Iterable`
  is the one we mention elsewhere for function parameters.  And as far
  as I can see, there isn't a reason we need to recommend `Iterator`
  instead here.

I'd have liked also to switch from recommending `six.text_type` to
`typing.Text`, but that should wait until 3.5.2, with `Text`, is out.

/cc @timabbott -- re-revisions welcome!

gvanrossum commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
-   # Mapping is an abtract base class for a dict-type thing that may
-   # not support writing to the mapping
+   # A generator function that yields ints is secretly just a function that
+   # returns an iterable (see below) of ints, so that's how we annotate it.
```
Technically it's an Iterator. Not sure if this distinction matters, given that Iterator inherits from Iterable.

gnprice replied:
Yeah. I wrote Iterator first for that reason, then contemplated explaining the distinction between Iterator and Iterable (having already recommended Iterable elsewhere), and concluded there isn't a good explanation in this context because the distinction doesn't actually matter.

gvanrossum commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
+           i += 1
+
+
+What to do when puzzled
```
I'm not sure Union and Any feel right under this heading.

gnprice replied:
Yeah. Perhaps "When you're puzzled or when things are complicated"? Or could group things differently.

gvanrossum replied:
This text review is getting pretty intense. Maybe better to commit and then
iterate?

gvanrossum commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
+   x = mystery_function() # type: Any
+
+   # Use `ignore` if you want to have something not be type-checked,
+   # to suppress mypy warnings for a given line.
```
IMO that's not strong enough. `ignore` is usually for suppressing complaints about things that confuse mypy or are outright bugs in mypy.

I think you need cast() in this section too.

It would also be nice to have something here about the situation where mypy complaints "Need type annotation for variable" (assignment to empty containers or None without a simple enough clue following, and None), and what to do about it.

timabbott replied:
yeah, good thoughts.  In Zulip, every `type: ignore` comes with a mypy bug number, and I think the recommendation should be closer to that...

gvanrossum commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
+
+   # typing.Match describes regex matches from the re module.
+   from typing import Match
+   x = re.match(r'[0-9]+', "15") # type: Match[str]
```
Also add typing.IO for streams. E.g. sys.stdout has type IO[str].

gvanrossum commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
-   # If something could be one of a few types, use Union
-   x = [3, 5, "test", "fun"] # type: List[Union[int, str]]
-   x = re.match(r'[0-9]+', "15") # type: Match[str]
+   # For textual data, we generally use six.text_type.
```
How should we introduce AnyStr? It's important but would be confusing this early.

gvanrossum commented:
Also, we really need to have a PY3 version of this...

gvanrossum commented:
Also looks like AppVeyor found something questionable about typed_ast. @ddfisher any idea? I think I'll just rename appveyor.yml until we have fixed this.

timabbott commented on docs/source/cheat_sheet.rst:
```diff
@@ -1,7 +1,7 @@
 ==================================
 
-This document is a quick cheat sheet showing how the PEP-484 type
+This document is a quick cheat sheet showing how the PEP 484 type
```
Make PEP 484 a link

timabbott commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
+   x = [3, 5, "test", "fun"] # type: List[Union[int, str]]
+
+   # Use Any if you don't know the type of something.
+   x = mystery_function() # type: Any
```
I think this probably needs more explanation of the tradeoffs around using Any.  E.g. for Zulip, we've been encouraging folks to only use Any in rare circumstances.

One thing I have noticed is `Dict[str, Any]` ends up being used a lot; might be worth mention that it's what you get out of 

`foo = dict(a=1, b='text', c=MyClass)`

ddfisher commented:
We now install typed_ast from PyPI as part of CI so we can run fast parser tests... and it doesn't build on Windows yet. :(

timabbott commented on docs/source/cheat_sheet.rst:
```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
+
+
+Standard duck types / ABCs
+**************************
```
you probably want to expand what those titles mean.  Most people writing Python code don't know those terms.

timabbott replied:
(Maybe with a link)

gnprice commented:
Thanks for all the review! Pushed a new version.

The comments that pointed out additional things we should cover, I added if they seemed pretty tricky to explain and otherwise I added as TODO items or will make issues for -- I want to start populating a "for newcomers" label on issues, composed mainly of easy fixes to documentation and the like. Lasse showed me how coala does this (see the issues at http://tinyurl.com/coala-new), and they've made very good use of it.

gnprice commented:
(Rebased to pick up Guido's disabling of AppVeyor.)

gvanrossum commented:
Somehow the appveyor.yml file still wreaked havoc; I've ignored it.
